src/java.base/share/classes/java/text/DigitList.java

```diff
@@ -508,94 +511,50 @@ private boolean shouldRoundUp(int maximumDigits,
-                            // Digit sequence either rounded up or truncated.
-                            // Round up only if it was truncated.
-                            return !alreadyRounded;
-                        }
-                    }
-                }
-                // Digit at rounding position is < '5' ==> no round up.
-                // Just let do the default, which is no round up (thus break).
-                break;
-            case HALF_EVEN:
-                // Implement IEEE half-even rounding
-                if (digits[maximumDigits] > '5') {
-                    return true;
-                } else if (digits[maximumDigits] == '5' ) {
-                    if (maximumDigits == (count - 1)) {
-                        // the rounding position is exactly the last index :
-                        if (alreadyRounded)
-                            // If FloatingDecimal rounded up (value was below tie),
-                            // then we should not round up again.
-                            return false;
-
-                        if (!valueExactAsDecimal)
-                            // Otherwise if the digits don't represent exact value,
-                            // value was above tie and FloatingDecimal truncated
-                            // digits to tie. We must round up.
-                            return true;
-                        else {
-                            // This is an exact tie value, and FloatingDecimal
-                            // provided all of the exact digits. We thus apply
-                            // HALF_EVEN rounding rule.
-                            return ((maximumDigits > 0) &&
-                                    (digits[maximumDigits-1] % 2 != 0));
-                        }
-                    } else {
-                        // Rounds up if it gives a non null digit after '5'
-                        for (int i=maximumDigits+1; i<count; ++i) {
-                            if (digits[i] != '0')
+                        // At tie, consider UP, DOWN, and EVEN logic
+                    } else if (digits[maximumDigits] == '5' ) {
+                        // Rounding position is the last index, there are 3 Cases.
+                        if (maximumDigits == (count - 1)) {
+                            // Do not round twice
+                            if (alreadyRounded) {
+                                return false;
+                                // When exact, consider specific contract logic
+                            } else if (valueExactAsDecimal) {
+                                return (roundingMode == RoundingMode.HALF_UP) ||
+                                        (roundingMode == RoundingMode.HALF_EVEN
+                                                && (maximumDigits > 0) && (digits[maximumDigits - 1] % 2 != 0));
+                                // Not already rounded, and not exact, round up
+                            } else {
```

Comment:
Are you sure this logic is equivalent to the previous one? Previously, `HALF_UP/DOWN` checks `valueExactAsDecimal` before `alreadyRounded`, but the new one checks `alreadyRounded` first.

Reply:
Yes, since `alreadyRounded` and `valueExactAsDecimal` are never both `true` and when either of them are `true`, it warrants a return without considering other logic. 

However, I have adjusted the code so that this is more apparent (and appears more similar to the original `HALF_DOWN/UP`, which was written more concisely).

Reply:
OK, thanks for confirming